.github/workflows/ci.yaml

```diff
@@ -22,42 +25,233 @@ permissions:
+
+      - name: Determine whether results will be 'published'
+        id: if-publish
         shell: bash -x -e {0}
         run: |
-          if [[ -z "${{ inputs. TRIAL_BRANCH}}" ]]; then
-            echo "TRIAL_BRANCH=${{ github.sha }}"
-          else
-            echo "TRIAL_BRANCH=${TRIAL_BRANCH}"
-          fi | tee $GITHUB_OUTPUT
+          echo "PUBLISH=${{ github.event_name == 'schedule' || (github.event_name == 'workflow_dispatch' && inputs.PUBLISH) }}" >> $GITHUB_OUTPUT
 
   amd64:
     needs: metadata
     uses: ./.github/workflows/_ci.yaml
     with:
       ARCHITECTURE: amd64
-      TRIAL_BRANCH: ${{ needs.metadata.outputs.TRIAL_BRANCH }}
     secrets: inherit
  
   arm64:
     needs: metadata
     uses: ./.github/workflows/_ci.yaml
     with:
       ARCHITECTURE: arm64
-      TRIAL_BRANCH: ${{ needs.metadata.outputs.TRIAL_BRANCH }}
     secrets: inherit
 
+  publish-mealkit:
+    needs: [metadata, amd64, arm64]
+    if: (!cancelled()) && needs.metadata.outputs.PUBLISH
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - TAG: JAX
+            TARGET_IMAGE: jax
+          - TAG: PALLAS
+            TARGET_IMAGE: pallas
+          - TAG: UPSTREAM_PAX
+            TARGET_IMAGE: upstream-pax
+          - TAG: UPSTREAM_T5X
+            TARGET_IMAGE: upstream-t5x
+          - TAG: PAX
+            TARGET_IMAGE: pax
+          - TAG: T5X
+            TARGET_IMAGE: t5x
+    uses: ./.github/workflows/_publish_container.yaml
+    with:
+      SOURCE_IMAGE: |
+        ${{ fromJson(needs.amd64.outputs.ALL_TAGS)[format('TAG_{0}_MEALKIT', matrix.TAG)] }}
+        ${{ fromJson(needs.arm64.outputs.ALL_TAGS)[format('TAG_{0}_MEALKIT', matrix.TAG)] }}
+      TARGET_IMAGE: ${{ matrix.TARGET_IMAGE }}
+      TARGET_TAGS: |
+        type=raw,value=mealkit,priority=500
+        type=raw,value=mealkit-${{ needs.metadata.outputs.BUILD_DATE }},priority=500
+
+  publish-final:
+    needs: [metadata, amd64, arm64]
+    if: (!cancelled()) && needs.metadata.outputs.PUBLISH
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - TAG: JAX
+            TARGET_IMAGE: jax
+          - TAG: PALLAS
+            TARGET_IMAGE: pallas
+          - TAG: UPSTREAM_PAX
+            TARGET_IMAGE: upstream-pax
+          - TAG: UPSTREAM_T5X
+            TARGET_IMAGE: upstream-t5x
+          - TAG: PAX
+            TARGET_IMAGE: pax
+          - TAG: T5X
+            TARGET_IMAGE: t5x
+    uses: ./.github/workflows/_publish_container.yaml
+    with:
+      SOURCE_IMAGE: |
+        ${{ fromJson(needs.amd64.outputs.ALL_TAGS)[format('TAG_{0}_FINAL', matrix.TAG)] }}
+        ${{ fromJson(needs.arm64.outputs.ALL_TAGS)[format('TAG_{0}_FINAL', matrix.TAG)] }}
+      TARGET_IMAGE: ${{ matrix.TARGET_IMAGE }}
+      TARGET_TAGS: |
+        type=raw,value=latest,priority=500
+        type=raw,value=nightly-${{ needs.metadata.outputs.BUILD_DATE }},priority=500
+
   finalize:
-    needs: [amd64, arm64]
+    runs-on: ubuntu-22.04
```

Comment:
Instead of breaking encapsulation and inlining the workflow here, how about we try to merge a few `_test_*.yaml` workflows?

Reply:
Reverted. Will do some investigation what can be merged to reduce the number of workflows.